main/src/main/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidator.java

```diff
@@ -41,27 +42,34 @@
 
   @Override
   public void validate(NoticeContainer noticeContainer) {
-    final Map<String, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
+    final Map<Integer, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
         new HashMap<>(fareRuleTable.entityCount());
     fareRuleTable
         .getEntities()
         .forEach(
             fareRule -> {
-              String fieldsCombination =
-                  fareRule.routeId()
-                      + fareRule.originId()
-                      + fareRule.containsId()
-                      + fareRule.destinationId();
-              if (fareRuleByZoneIdFieldsCombination.containsKey(fieldsCombination)) {
+              GtfsFareRule otherFareRule =
+                  fareRuleByZoneIdFieldsCombination.putIfAbsent(getHash(fareRule), fareRule);
+              if (otherFareRule != null) {
                 noticeContainer.addValidationNotice(
                     new DuplicateFareRuleZoneIdFieldsNotice(
                         fareRule.csvRowNumber(),
                         fareRule.fareId(),
-                        fareRuleByZoneIdFieldsCombination.get(fieldsCombination).csvRowNumber(),
-                        fareRuleByZoneIdFieldsCombination.get(fieldsCombination).fareId()));
-              } else {
-                fareRuleByZoneIdFieldsCombination.put(fieldsCombination, fareRule);
+                        otherFareRule.csvRowNumber(),
+                        otherFareRule.fareId()));
               }
             });
   }
+
+  /**
+   * Returns the hashcode associated to the combination of "fare_rules.origin_id",
+   * "fare_rules.contains_id" and "fare_rules.destination_id".
+   *
+   * @param fareRule the {@code GtfsFareRule} to generate the hash from
+   * @return the hashcode associated to the combination of this {@code GtfsFareRule} "origin_id",
+   *     "contains_id" and "destination_id".
+   */
+  private int getHash(GtfsFareRule fareRule) {
```

Comment:
As discussed `route_id` should be part of this composite key (https://github.com/MobilityData/gtfs-validator/pull/713#issuecomment-776047895). Opening an issue to include this.